### PR TITLE
[Fix] Support sparse optimizer DCP resume

### PR DESCRIPTION
**Pull request by matrix72c**
## Motivation

PyTorch optimizers initialize per-parameter state lazily. A valid DCP checkpoint can therefore omit all optimizer-state leaves for a trainable parameter that did not receive a gradient before save. During resume, XTuner materializes a full optimizer load target, and DCP reports those absent lazy leaves as missing checkpoint keys.

## Changes

- Inspect DCP metadata before loading optimizer state and remove only parameter states whose leaves are wholly absent from the checkpoint.
- Keep partially missing optimizer states fatal so corrupted AdamW state is not silently accepted.
- Preserve model strictness while loading sparse optimizer state with the PyTorch 2.10+ `strict=False` behavior.
- Add a regression test covering AdamW DCP resume, later lazy-state materialization, and rejection of partially saved state.

## Impact

Model checkpoint validation remains unchanged. Existing optimizer states are restored normally; only wholly absent lazy per-parameter state is deferred until that parameter next receives a gradient.

## Tests

- XTuner pre-commit hooks passed for the changed files, except mypy (the hook environment exhausted local temporary storage while installing its pinned PyTorch 2.6 CUDA dependencies).
- `ruff check xtuner/v1/engine/train_engine.py tests/engine/test_train_engine_checkpoint.py`
- `python3.12 -m py_compile xtuner/v1/engine/train_engine.py tests/engine/test_train_engine_checkpoint.py`
- PyTorch 2.10 DCP save/load/optimizer-step reproduction passed.

### CODE REVIEW

**Comment by YanhuiDua** [comment]
@claude review

File: xtuner/v1/engine/train_engine.py [hunk 1]
```diff
@@ -497,15 +538,22 @@ def load_dcp(
         load_optimizer = load_states or load_args
         state_dict = self._get_dcp_state_dict(cpu_offload=True, save_optimizer=load_optimizer)
 
-        if self.has_freeze_params:
-            set_options = StateDictOptions(cpu_offload=True, strict=False)
-        else:
-            set_options = StateDictOptions(cpu_offload=True, strict=True)
+        model_options = StateDictOptions(cpu_offload=True, strict=not self.has_freeze_params)
```
Comment (YanhuiDua) [inline xtuner/v1/engine/train_engine.py:541]: 这里为啥要 set_options 改为 model_options

File: xtuner/v1/engine/train_engine.py [hunk 2]
```diff
@@ -497,15 +538,22 @@ def load_dcp(
...
 
         with profile_time_and_memory(f"[Load DCP from {weights_dir}]"):
+            if load_optimizer:
+                metadata = dcp.FileSystemReader(weights_dir).read_metadata()
+                saved_keys = {str(key) for key in metadata.state_dict_metadata}
+                removed = _prune_uncheckpointed_optimizer_state(state_dict["optimizer"], saved_keys)
```
[the remark this answers is not in the document]
Comment (YanhuiDua) [reply xtuner/v1/engine/train_engine.py:547]: 这个问题需要修复

File: tests/engine/test_train_engine_checkpoint.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        }
+
+        dcp.load(load_target, checkpoint_id=checkpoint_dir)
+        set_model_state_dict(restored_model, load_target["model"])
+        set_optimizer_state_dict(
+            restored_model,
+            restored_optimizer,
+            optim_state_dict=load_target["optimizer"],
+            options=StateDictOptions(strict=False),
+        )
```
[the remark this answers is not in the document]
Comment (YanhuiDua) [reply tests/engine/test_train_engine_checkpoint.py:60]: 这个也需要修复下

File: xtuner/v1/engine/train_engine.py [hunk 4]
```diff
@@ -58,6 +58,47 @@ class TrainStepInfo(DataBatchInfo, BatchForwardInfo):
...
+    ``get_optimizer_state_dict`` materializes optimizer states for every
+    trainable parameter. A checkpoint may legitimately omit all state slots
+    for a parameter that had not received a gradient when it was saved. DCP
+    treats the materialized load target as authoritative and otherwise reports
+    those absent slots as missing checkpoint keys.
+
+    A partially saved parameter state is not lazy initialization and is kept as
+    an error: optimizers such as AdamW cannot step with only a subset of
+    ``step``, ``exp_avg``, and ``exp_avg_sq`` restored.
+    """
```
[the remark this answers is not in the document]
Comment (YanhuiDua) [reply xtuner/v1/engine/train_engine.py:74]: 建议作为trainengine的私有方法

File: xtuner/v1/engine/train_engine.py [hunk 5]
```diff
@@ -532,7 +580,9 @@ def load_dcp(
                 self.model,
                 self.optimizer,
                 optim_state_dict=optimizer_state_dict,
-                options=set_options,
+                # PyTorch 2.10+ treats wholly absent per-parameter optimizer
+                # states as valid lazy state when strict loading is disabled.
+                options=StateDictOptions(cpu_offload=True, strict=False),
```
[the remark this answers is not in the document]
Comment (YanhuiDua) [reply xtuner/v1/engine/train_engine.py:585]: 这里是不是不要直接给默认参数为False
